libs/libcommon/tests/test_smart_update.py

```diff
@@ -0,0 +1,283 @@
+"""
+
+
+ADD_TAG_DIFF = """
+diff --git a/README.md b/README.md
+new file mode 100644
+index 6ea47cd9c61754a4838b07f479b376cb431f2271..6ea47cd9c61754a4838b07f479b376cb431f2273
+--- a/README.md
++++ b/README.md
+@@ -0,0 +5,6 @@
++tags:
++- test
+"""
+
+ADD_DATA_DIFF = """
+diff --git a/data.csv b/data.csv
+new file mode 100644
+index 6ea47cd9c61754a4838b07f479b376cb431f2271..6ea47cd9c61754a4838b07f479b376cb431f2274
+--- a/README.md
++++ b/data.csv
+@@ -0,0 +1,3 @@
++id,text
++0,foo
++1,bar
+"""
+
+INITIAL_README = f"# Dataset Card for {DATASET_NAME}"
+ONE_CONFIG_README = (
+    f"---\nconfigs:\n- config_name: foo\n  data_files: foo.csv\n---\n\n# Dataset Card for {DATASET_NAME}"
+)
+TWO_CONFIGS_README = f"---\nconfigs:\n- config_name: foo\n  data_files: foo.csv\n- config_name: bar\n  data_files: bar.csv\n---\n\n# Dataset Card for {DATASET_NAME}"
+ONE_CONFIG_AND_ONE_TAG_README = f"---\nconfigs:\n- config_name: foo\n  data_files: foo.csv\ntags:\n- test\n---\n\n# Dataset Card for {DATASET_NAME}"
+
+
+@pytest.fixture(autouse=True)
+def queue_mongo_resource_autouse(queue_mongo_resource: QueueMongoResource) -> QueueMongoResource:
+    return queue_mongo_resource
+
+
+@pytest.fixture(autouse=True)
+def cache_mongo_resource_autouse(cache_mongo_resource: CacheMongoResource) -> CacheMongoResource:
+    return cache_mongo_resource
+
+
+def test_empty_cache() -> None:
+    # No cache, raise
+    with pytest.raises(SmartUpdateImpossibleBecauseCacheIsEmpty):
+        get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+
+
+def test_same_revision() -> None:
+    put_cache(step=STEP_DA, dataset=DATASET_NAME, revision=REVISION_NAME)
+    # Same revision as cache, do nothing
+    plan = get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+    assert_smart_dataset_update_plan(plan, tasks=[])
+
+
+def test_cache_revision_is_not_parent_revision_commit() -> None:
+    # If rewriting git history OR spamming commits OR out of order commits OR in case of concurrency issues: raise
+    put_cache(step=STEP_DA, dataset=DATASET_NAME, revision="not_parent_revision")
+    with pytest.raises(SmartUpdateImpossibleBecauseCachedRevisionIsNotParentOfNewRevision):
+        get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+
+
+def test_empty_commit() -> None:
+    # Empty commit: update the revision of the cache entries
+    put_cache(step=STEP_DA, dataset=DATASET_NAME, revision=OTHER_REVISION_NAME)
+    with put_diff(EMPTY_DIFF):
+        plan = get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+        assert_smart_dataset_update_plan(
+            plan,
+            files_impacted_by_commit=[],
+            tasks=["UpdateRevisionOfDatasetCacheEntriesTask,1"],
+        )
+    with pytest.raises(RequestException):  # if diff doesn't exist
+        get_smart_dataset_update_plan(processing_graph=PROCESSING_GRAPH_TWO_STEPS)
+
+
+def test_add_initial_readme_commit() -> None:
```

Comment:
another interesting test would be if the commit creates the README file, with the `configs` field in YAML for example -> it should raise `SmartUpdateImpossibleBecauseOfUpdatedYAMLField` iirc